src/Cronofy/AddToCalendarRequestBuilder.cs

```diff
@@ -0,0 +1,146 @@
+﻿namespace Cronofy
+{
+    using System;
+    using Cronofy.Requests;
+
+    /// <summary>
+    /// Class to build an add to calendar request.
+    /// </summary>
+    public sealed class AddToCalendarRequestBuilder : IBuilder<AddToCalendarRequest>
+    {
+        /// <summary>
+        /// The oauth details for the request.
+        /// </summary>
+        private IBuilder<AddToCalendarRequest.OAuthDetails> oauthBuilder;
+
+        /// <summary>
+        /// The event details builder for the request.
+        /// </summary>
+        private IBuilder<UpsertEventRequest> upsertEventRequestBuilder;
+
+        /// <summary>
+        /// Sets the OAuth details of the request.
+        /// </summary>
+        /// <param name="redirectUri">
+        /// The redirect uri for the request's oauth details, must not be blank.
+        /// </param>
+        /// <param name="scope">
+        /// The scope for the request's oauth details, must not be blank.
+        /// </param>
+        /// <returns>
+        /// A reference to the <see cref="AddToCalendarRequestBuilder"/>.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// Thrown if <paramref name="redirectUrl"/> is empty.  
+        /// </exception>
+        public AddToCalendarRequestBuilder OAuthDetails(string redirectUri, string scope)
+        {
+            return this.OAuthDetails(redirectUri, scope, null);
+        }
+
+        /// <summary>
+        /// Sets the OAuth details of the request.
+        /// </summary>
+        /// <param name="redirectUri">
+        /// The redirect uri for the request's oauth details, must not be blank.
+        /// </param>
+        /// <param name="scope">
+        /// The scope for the request's oauth details, must not be blank.
+        /// </param>
+        /// <param name="state">
+        /// The state for the request's oauth details.
+        /// </param>
+        /// <returns>
+        /// A reference to the <see cref="AddToCalendarRequestBuilder"/>.
+        /// </returns>
+        /// <exception cref="ArgumentException">
+        /// Thrown if <paramref name="redirectUrl"/> or <paramref name="scope"/> are empty.  
+        /// </exception>
+        public AddToCalendarRequestBuilder OAuthDetails(string redirectUri, string scope, string state)
+        {
+            Preconditions.NotBlank("redirectUri", redirectUri);
+
+            if (scope != null)
+            {
+                Preconditions.NotBlank("scope", scope);
+            }
+
+            this.oauthBuilder = new BuilderWrapper<AddToCalendarRequest.OAuthDetails>(
```

Comment:
These are a bit ugly, this instance particularly. Some extra gymnastics around the wrapper will tidy things up at the multiple places you call this.

```cs
internal static class BuilderWrapper
{
    private sealed class Wrapper<T> : IBuilder<T>
    {
        ctor(instance)

        public T Build()
        {
            return this.instance;
        }
    }
    public IBuilder<T> For<T>(T instance)
    {
        return new Wrapper<T>(instance);
    }
}
```

Then rather than what you have here, you end up with:

```cs
var oauthDetails = new AddToCalendarRequest.OAuthDetails
    {
        RedirectUri = redirectUri,
        Scope = scope,
        State = state
    };

this.upsertEventRequestBuilder = BuilderWrapper.For(oauthDetails);
```